src/Microsoft.AspNet.OData.Shared/Formatter/Serialization/ODataResourceValueSerializer.cs

```diff
@@ -0,0 +1,156 @@
+            : base(payloadKind, serializerProvider)
+        {
+            if (serializerProvider == null)
+            {
+                throw Error.ArgumentNull("serializerProvider");
+            }
+          
+        }
+
+        /// <inheritdoc/>
+        public override void WriteObject(object graph, Type type, ODataMessageWriter messageWriter, ODataSerializerContext writeContext)
+        {
+            if (messageWriter == null)
+            {
+                throw Error.ArgumentNull("messageWriter");
+            }
+            if (writeContext == null)
+            {
+                throw Error.ArgumentNull("writeContext");
+            }
+            if (writeContext.RootElementName == null)
+            {
+                throw Error.Argument("writeContext", SRResources.RootElementNameMissing, typeof(ODataSerializerContext).Name);
+            }
+
+            IEdmTypeReference edmType = writeContext.GetEdmType(graph, type);
+            Contract.Assert(edmType != null);
+
+            messageWriter.WriteProperty(CreateProperty(graph, edmType, writeContext.RootElementName, writeContext));
+        }
+
+        /// <inheritdoc/>
+        public sealed override ODataValue CreateODataValue(object graph, IEdmTypeReference expectedType, ODataSerializerContext writeContext)
+        {
+            if (!expectedType.IsStructured())
+            {
+                throw Error.InvalidOperation(SRResources.CannotWriteType, typeof(ODataResourceValueSerializer), expectedType.FullName());
+            }
+
+            ODataResourceValue value = CreateODataResourceValue(graph, expectedType.AsStructured(), writeContext);
+            if (value == null)
+            {
+                return new ODataNullValue();
+            }
+
+            return value;
+        }
+
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Performance", "CA1804:RemoveUnusedLocals", MessageId = "edmTypeSerializer")]
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Performance", "CA1822:MarkMembersAsStatic")]
+        private ODataResourceValue CreateODataResourceValue(object graph, IEdmStructuredTypeReference expectedType, ODataSerializerContext writeContext)
+        {
+            List<ODataProperty> properties = new List<ODataProperty>();
+            ODataResourceValue resourceValue = new ODataResourceValue();
+
+            IDelta delta = graph as IDelta;
+            if (delta != null)
+            {
+                foreach (string propertyName in delta.GetChangedPropertyNames())
+                {
+                    object propertyValue;
+                    delta.TryGetPropertyValue(propertyName, out propertyValue);
+
+                    if (propertyValue != null)
+                    {
+                        SetPropertyValue(writeContext, properties, propertyName, propertyValue);
+                    }
```

Comment:
The property may intentionally be null.  Rather than testing on whether propertyValue is null, I would check the return value from TryGetPropertyValue:

                    if (delta.TryGetPropertyValue(propertyName, out propertyValue)
                    {
                        SetPropertyValue(writeContext, properties, propertyName, propertyValue);
                    }